File: client/src/lib/review.ts
/**
 * Match-review data layer (spec §4.3 FR-L4/FR-L5/FR-L6, roadmap M5 and M13).
 *
 * A file the matcher could not link with enough confidence goes to review
 * rather than being auto-linked to a guess, and nothing is played from it
 * until a person says which show it is. This module is everything the Review
 * page and the sidebar badge need to work that queue: the queue itself, the
 * three ways out of it (confirm, ignore, reopen), a catalogue search hung off
 * an item for the show the matcher never proposed, and the request for an LLM
 * suggestion — which is *shown*, never applied (FR-L5).
 *
 * Every mutation invalidates both the queue and the summary count, because
 * both are server arithmetic: a confirm removes a row and drops the badge, an
 * ignore does the same, a reopen puts one back. Nothing here patches a cache
 * optimistically — the queue is small, the round trip is cheap, and a badge
 * that briefly disagrees with the list is worse than one that lags by 200 ms.
 *
 * Confirming or ignoring also moves an *episode* (`matched`, or back to
 * `unavailable` when Arc's own download turns out to be the wrong file), so
 * the anime tree is invalidated with them; a show page left open must not go
 * on claiming the old state.
 */

import { useEffect, useState } from 'react'
import {
  keepPreviousData,
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from '@tanstack/react-query'
import { ApiError, apiFetch } from '@/lib/api'
import {
  ANIME_QUERY_KEY,
  CATALOGUE_UNAVAILABLE_MESSAGE,
  isSearchable,
  type AnimeSearchResponse,
  type AnimeSummary,
} from '@/lib/anime'
import { errorDetail, useMe } from '@/lib/auth'

/* --- Wire shapes ------------------------------------------------------ */

/** The matcher's verdict on a file, as `arc.models.enums.ReviewState`. */
export type ReviewState = 'auto' | 'pending' | 'confirmed' | 'ignored'

/** The three states the queue can be listed in; `confirmed` has its own home. */
export const REVIEW_TABS = ['pending', 'ignored', 'auto'] as const
export type ReviewTab = (typeof REVIEW_TABS)[number]

export const REVIEW_TAB_LABELS: Record<ReviewTab, string> = {
  pending: 'Pending',
  ignored: 'Ignored',
  auto: 'Auto-linked',
}

/** The part of the filename parse the queue renders (server `ParsedOut`). */
export interface ReviewParsed {
  title: string | null
  episode: number | null
  season: number | null
  group: string | null
  resolution: string | null
  kind: string | null
}

/**
 * One scored candidate. `anime` is null — and `reason` set instead — when the
 * matcher had nothing to offer and the entry is a sentence rather than a show.
 */
export interface ReviewCandidate {
  anime: AnimeSummary | null
  episode_number: number | null
  score: number | null
  reasons: string[]
  /** True when the episode number came from the absolute-numbering rule. */
  absolute: boolean
  /** Set instead of `anime`: "no good candidates". */
  reason: string | null
}

/** How sure the model says it is. Never a number: it did not measure one. */
export type SuggestionConfidence = 'high' | 'medium' | 'low'

/**
 * What a language model proposed for a file it was asked about (FR-L5).
 *
 * Never applied by anything: the page renders it, and a person presses
 * Confirm — or does not. `error` is set instead of a proposal when the call
 * failed or the model declined, so a failed ask still has something to say.
 */
export interface ReviewSuggestion {
  anime_id: number | null
  anime: AnimeSummary | null
  episode_number: number | null
  /**
   * One line of argument, in the model's own words. Null exactly when `error`
   * is set: an ask that failed has an explanation, not a case.
   */
  reason: string | null
  /** Null on the same terms as `reason`: nothing was proposed to be sure of. */
  confidence: SuggestionConfidence | null
  /** The model that answered, named verbatim. Null for an older row. */
  model: string | null
  created_at: string
  /** Why there is no proposal, when there is none. */
  error: string | null
}

/** One row of the queue (server `ReviewItem`). Never carries a path. */
export interface ReviewItem {
  id: number
  /** The filename alone. */
  name: string
  /** Its directory relative to `DATA_DIR`; `""` for the library root. */
  directory: string
  size: number | null
  parsed: ReviewParsed
  /** The matcher's score, 0–1. Null once a person has decided. */
  confidence: number | null
  candidates: ReviewCandidate[]
  review_state: ReviewState
  /** The episode this file is linked to, once it is. Null while pending. */
  episode_id: number | null
  created_at: string
  /**
   * Optional on the wire: a server without the suggestion job simply omits
   * the key, and `suggestionOf` is how every reader should get at it so a
   * missing key and an explicit null stay the same thing.
   */
  suggestion?: ReviewSuggestion | null
}

/** `GET /api/review` — the queue, and how much of it is pending. */
export interface ReviewPage {
  items: ReviewItem[]
  /**
   * Always the *pending* count, whatever `?state=` asked for: the sidebar
   * badge renders this and must not change because somebody filtered.
   */
  pending: number
  /**
   * False when the server has no model key or `LLM_MATCH_SUGGESTIONS` is off.
   * Optional for the same reason `suggestion` is; read it through
   * `suggestionsEnabled`.
   */
  suggestions_enabled?: boolean
}

/** `GET /api/review/summary` — the one number the sidebar polls. */
export interface ReviewSummary {
  /** Media files waiting for this viewer to resolve them. */
  pending: number
}

/** `POST /api/review/{id}/confirm` (FR-L6). Both fields are set by hand. */
export interface ConfirmReview {
  id: number
  anime_id: number
  episode_number: number
}

/** `POST /api/review/{id}/suggest` — 202, because a job does the asking. */
export interface SuggestionRequested {
  job_id: number
  status: string
}

/** An item's suggestion, with a server that has none reading as "none". */
export function suggestionOf(item: ReviewItem): ReviewSuggestion | null {
  return item.suggestion ?? null
}

/** Whether this server will ask a model at all (FR-L5's toggle). */
export function suggestionsEnabled(page: ReviewPage): boolean {
  return page.suggestions_enabled ?? false
}

/* --- Query keys ------------------------------------------------------- */

export const REVIEW_QUERY_KEY = 'review'

export const reviewSummaryQueryKey = [REVIEW_QUERY_KEY, 'summary'] as const

/** Every listing of the queue, under one prefix so all of them invalidate. */
export const reviewQueuesQueryKey = [REVIEW_QUERY_KEY, 'queue'] as const

export function reviewQueueQueryKey(state: ReviewTab): readonly unknown[] {
  return [...reviewQueuesQueryKey, state]
}

export function reviewSearchQueryKey(id: number, q: string): readonly unknown[] {
  return [REVIEW_QUERY_KEY, 'search', id, q]
}

/* --- Timings ---------------------------------------------------------- */

/** How long a fetched count is treated as fresh. */
const REVIEW_STALE_TIME_MS = 60_000

/** How often the count is refetched while the app is open. */
const REVIEW_REFETCH_INTERVAL_MS = 120_000

/** Long enough that a typed word is one request, short enough to feel live. */
export const SEARCH_DEBOUNCE_MS = 300

/** How often the queue is re-asked while a suggestion is being generated. */
export const SUGGESTION_POLL_MS = 10_000

/**
 * How long to keep asking. A model call takes seconds; two minutes is the
 * point at which something has gone wrong on the server and polling is only
 * noise. The page still has a Try-again button after it.
 */
export const SUGGESTION_WAIT_MS = 120_000

/* --- Words ------------------------------------------------------------ */

/** The 409 detail that means "somebody linked this while you were looking". */
const ALREADY_LINKED_DETAIL = 'this file is already linked to an episode'

/** The 409 detail `reopen` answers when the file is not ignored any more. */
const NOT_IGNORED_DETAIL = 'only an ignored file can be reopened'

/** The 503 detail that means the server will not ask a model at all. */
const SUGGESTIONS_OFF_DETAIL = 'Suggestions are not enabled'

export const ALREADY_LINKED_MESSAGE =
  'That file is already linked to an episode. Refresh the queue to see where it went.'

export const GONE_MESSAGE = 'That file is no longer in the queue. Refresh to see the current list.'

export const NOT_IGNORED_MESSAGE =
  'That file is not ignored any more. Refresh the queue to see its current state.'

export const SUGGESTIONS_OFF_MESSAGE =
  'Suggestions are switched off on this server. An administrator can enable them.'

/**
 * What to say when a review action fails. Every case is something the person
 * can act on — refresh, wait, ask an admin — except the last, which is at
 * least honest about not knowing.
 */
export function reviewErrorMessage(error: unknown): string {
  if (!(error instanceof ApiError)) return 'Could not reach the server. Try again.'

  const detail = errorDetail(error)
  switch (error.status) {
    case 401:
      return 'Your session expired. Sign in again.'
    case 404:
      return GONE_MESSAGE
    case 409:
      return detail === NOT_IGNORED_DETAIL ? NOT_IGNORED_MESSAGE : ALREADY_LINKED_MESSAGE
    case 502:
      return CATALOGUE_UNAVAILABLE_MESSAGE
    case 503:
      return detail === SUGGESTIONS_OFF_DETAIL ? SUGGESTIONS_OFF_MESSAGE : 'Try again in a moment.'
    default:
      return 'Something went wrong. Try again.'
  }
}

/** Kept so a caller can tell the two 409s apart without re-deriving them. */
export const REVIEW_CONFLICT_DETAILS = {
  alreadyLinked: ALREADY_LINKED_DETAIL,
  notIgnored: NOT_IGNORED_DETAIL,
} as const

/* --- Formatting ------------------------------------------------------- */

const SIZE_UNITS = ['B', 'KB', 'MB', 'GB', 'TB'] as const

/**
 * "1.4 GB". The exact byte count is never the question — "is this the whole
 * episode or a 40 kB sample?" is — so one decimal below a hundred is plenty.
 */
export function formatSize(bytes: number | null): string {
  if (bytes === null || !Number.isFinite(bytes) || bytes < 0) return 'size unknown'

  let value = bytes
  let unit = 0
  while (value >= 1024 && unit < SIZE_UNITS.length - 1) {
    value /= 1024
    unit += 1
  }
  const rounded = unit === 0 || value >= 100 ? Math.round(value) : Number(value.toFixed(1))
  return `${String(rounded)} ${SIZE_UNITS[unit] ?? 'B'}`
}

/**
 * The matcher's score as a percentage. A file whose score has been cleared —
 * which is what confirming does — has nothing to report rather than "0 %".
 */
export function formatConfidence(confidence: number | null): string {
  if (confidence === null || !Number.isFinite(confidence)) return 'no score'
  return `${String(Math.round(confidence * 100))}% sure`
}

/** A candidate's score as a bare percentage; blank when it has none. */
export function formatScore(score: number | null): string {
  if (score === null || !Number.isFinite(score)) return ''
  return `${String(Math.round(score * 100))}%`
}

/** "S2E05", "E05", "movie" — the episode as the parse saw it, or ''. */
export function episodeLabel(parsed: ReviewParsed): string {
  if (parsed.episode === null) return ''
  const episode = `E${String(parsed.episode).padStart(2, '0')}`
  return parsed.season === null ? episode : `S${String(parsed.season)}${episode}`
}

/**
 * The parse as a handful of chips: what the filename claimed, in the order a
 * person reads it. Anything the parser could not find is simply absent — an
 * empty chip says less than no chip.
 */
export function parsedChips(parsed: ReviewParsed): string[] {
  const chips: string[] = []
  if (parsed.title !== null && parsed.title !== '') chips.push(parsed.title)
  const episode = episodeLabel(parsed)
  if (episode !== '') chips.push(episode)
  if (parsed.group !== null && parsed.group !== '') chips.push(parsed.group)
  if (parsed.resolution !== null && parsed.resolution !== '') chips.push(parsed.resolution)
  if (parsed.kind !== null && parsed.kind !== '' && parsed.kind !== 'episode') {
    chips.push(parsed.kind)
  }
  return chips
}

/** Where the file sits, in words. The root of the library has a name here. */
export function directoryLabel(directory: string): string {
  return directory === '' ? 'library root' : directory
}

/**
 * The episode number to start the confirm form at: the candidate's own, else
 * whatever the filename claimed, else nothing for a person to fill in.
 */
export function startingEpisode(
  candidateEpisode: number | null,
  parsed: ReviewParsed,
): number | null {
  return candidateEpisode ?? parsed.episode
}

/** An episode number a person typed, or null when it is not one. */
export function parseEpisodeInput(raw: string): number | null {
  const trimmed = raw.trim()
  if (trimmed === '') return null
  const value = Number(trimmed)
  return Number.isInteger(value) && value >= 1 ? value : null
}

/* --- Queries ---------------------------------------------------------- */

/**
 * The pending-review count for the signed-in viewer.
 *
 * The count moves when the ingest job runs, not when the viewer clicks, so it
 * is polled on a slow interval instead of invalidated. `retry: false` because
 * a badge is not worth a retry storm. The endpoint needs a session, so the
 * query stays disabled until `me` has answered: firing it while logged out
 * would only ever earn a 401, which the query client reads as the session
 * dying under us.
 */
export function useReviewSummary(): UseQueryResult<ReviewSummary, Error> {
  const { data: me } = useMe()

  return useQuery<ReviewSummary, Error>({
    queryKey: reviewSummaryQueryKey,
    queryFn: () => apiFetch<ReviewSummary>('/api/review/summary'),
    staleTime: REVIEW_STALE_TIME_MS,
    refetchInterval: REVIEW_REFETCH_INTERVAL_MS,
    retry: false,
    enabled: me != null,
  })
}

/**
 * One state's worth of the queue.
 *
 * `pollUntil` is an epoch millisecond deadline, set while a suggestion has
 * been asked for and has not arrived: the answer lands in a job, so the only
 * way to see it is to re-ask. Reading the clock inside the interval callback
 * rather than closing over a boolean is what makes the deadline self-ending —
 * the first tick past it returns `false` and the polling stops with no timer
 * of its own.
 */
export function useReviewQueue(
  state: ReviewTab,
  pollUntil: number | null = null,
): UseQueryResult<ReviewPage, Error> {
  return useQuery<ReviewPage, Error>({
    queryKey: reviewQueueQueryKey(state),
    queryFn: () => apiFetch<ReviewPage>(`/api/review?state=${state}`),
    retry: false,
    // Switching tabs must not blank the list under the tabs that did it.
    placeholderData: keepPreviousData,
    refetchInterval: () =>
      pollUntil !== null && Date.now() < pollUntil ? SUGGESTION_POLL_MS : false,
  })
}

/**
 * `value`, but only once it has stopped changing for `delayMs`.
 *
 * Here rather than in the page because the debounce belongs to the request,
 * not to the input: every caller of `useReviewSearch` should cost one call per
 * typed word, and none of them should have to remember to arrange it.
 */
export function useDebounced<T>(value: T, delayMs: number = SEARCH_DEBOUNCE_MS): T {
  const [settled, setSettled] = useState(value)

  useEffect(() => {
    if (settled === value) return
    const timer = setTimeout(() => {
      setSettled(value)
    }, delayMs)
    return () => {
      clearTimeout(timer)
    }
  }, [value, settled, delayMs])

  return settled
}

/**
 * Catalogue search for the show the matcher never proposed (FR-L6).
 *
 * Hung off the item's id, as the server has it, so the client does not have to
 * hold two ideas at once. Each result is upserted on the way past, which is
 * what gives it the internal id `confirm` then takes. `retry: false`: every
 * attempt costs the server a catalogue call, and neither a rejected query nor
 * a failing upstream gets better for being asked twice.
 */
export function useReviewSearch(id: number, q: string): UseQueryResult<AnimeSearchResponse, Error> {
  const query = useDebounced(q.trim())

  return useQuery<AnimeSearchResponse, Error>({
    queryKey: reviewSearchQueryKey(id, query),
    queryFn: () => {
      const params = new URLSearchParams({ q: query })
      return apiFetch<AnimeSearchResponse>(`/api/review/${String(id)}/search?${params.toString()}`)
    },
    enabled: Number.isInteger(id) && id > 0 && isSearchable(query),
    placeholderData: keepPreviousData,
    retry: false,
  })
}

/* --- Mutations -------------------------------------------------------- */

/**
 * Every action here changes the queue, the badge, and — because confirming
 * and ignoring both move an episode — what a show page says about its own
 * state. All three are server-derived, so all three are re-asked rather than
 * patched.
 */
function useReviewMutation<TVariables>(
  request: (variables: TVariables) => Promise<ReviewItem>,
): UseMutationResult<ReviewItem, Error, TVariables> {
  const client = useQueryClient()

  return useMutation<ReviewItem, Error, TVariables>({
    mutationFn: request,
    onSuccess: () => {
      void client.invalidateQueries({ queryKey: reviewQueuesQueryKey })
      void client.invalidateQueries({ queryKey: reviewSummaryQueryKey })
      void client.invalidateQueries({ queryKey: [ANIME_QUERY_KEY] })
    },
  })
}

/** Link a file to an episode by hand (FR-L6). 409 when it is already linked. */
export function useConfirmReview(): UseMutationResult<ReviewItem, Error, ConfirmReview> {
  return useReviewMutation<ConfirmReview>(({ id, anime_id, episode_number }) =>
    apiFetch<ReviewItem>(`/api/review/${String(id)}/confirm`, {
      method: 'POST',
      body: JSON.stringify({ anime_id, episode_number }),
    }),
  )
}

/** "Not anime / ignore" (FR-L6): out of the queue, still on disk. */
export function useIgnoreReview(): UseMutationResult<ReviewItem, Error, number> {
  return useReviewMutation<number>((id) =>
    apiFetch<ReviewItem>(`/api/review/${String(id)}/ignore`, { method: 'POST' }),
  )
}

/** Put an ignored file back in the queue. Only from `ignored`, hence the 409. */
export function useReopenReview(): UseMutationResult<ReviewItem, Error, number> {
  return useReviewMutation<number>((id) =>
    apiFetch<ReviewItem>(`/api/review/${String(id)}/reopen`, { method: 'POST' }),
  )
}

/**
 * Ask a model what this file probably is (FR-L5). The answer arrives in a
 * job, so the 202 says only that the asking has started; the queue is
 * invalidated with it for consistency with the other actions, and the page
 * polls until the suggestion appears.
 *
 * Nothing this returns is ever applied: the suggestion is rendered, and a
 * person presses Confirm.
 */
export function useRequestSuggestion(): UseMutationResult<SuggestionRequested, Error, number> {
  const client = useQueryClient()

  return useMutation<SuggestionRequested, Error, number>({
    mutationFn: (id) =>
      apiFetch<SuggestionRequested>(`/api/review/${String(id)}/suggest`, { method: 'POST' }),
    onSuccess: () => {
      void client.invalidateQueries({ queryKey: reviewQueuesQueryKey })
      void client.invalidateQueries({ queryKey: reviewSummaryQueryKey })
    },
  })
}

File: server/arc/api/review.py
"""The match-review queue over HTTP (FR-L4, FR-L6, FR-D4).

Phase 1 is API only; the page lands in M13. The contract is already the one the
page needs: list, a cheap count for the sidebar badge, confirm, ignore, reopen,
and a search so somebody can pick a title the matcher never proposed.

**Who may use it.** Any signed-in user, not just an admin. spec §2 gives an
ordinary user "resolve match-review items for their own requested shows", and
in a single-library server every file is potentially somebody's — there is no
per-user ownership on ``media_files`` to filter by. Admin-only would mean the
owner resolves everybody's queue, which is the opposite of the intent. The
admin-wide view (FR-D4) is this same endpoint.

**What it never returns.** An absolute path. See
:mod:`arc.api.review_schemas`.

**What confirming does.** Exactly what an automatic link does, through the same
function (:func:`arc.services.library.link.link`), with ``review_state`` set to
``confirmed`` instead of ``auto``, and with ``match_confidence`` set back to
null: the number described how sure the *matcher* was, and once a person has
chosen the show there is nothing left for it to describe. The candidate list
is kept — it is what was on screen when the choice was made.

**What ignoring does.** Takes the file out of the queue — and, if Arc
downloaded it, tells the *episode* so: the release Arc chose was not this
episode after all, so the episode leaves ``matching`` for ``unavailable`` and
the daily retry looks again (:mod:`arc.services.acquisition.reject`). A file
somebody dropped in by hand carries no such claim and changes nothing.

A file that is already linked is a 409, not a silent relink: the client should
reopen it first, so that "I changed my mind" is a deliberate two-step rather
than a double-click. Since every linked state (``auto``, ``confirmed``) carries
an ``episode_id`` and no unlinked one does, that check is also, exactly, "only
a ``pending`` or ``ignored`` file may be confirmed" — which is the rule as a
person would state it. Confirming an ``ignored`` file is deliberate and
allowed: "not anime after all — no, wait, it is *this*" needs no detour
through ``reopen``.
"""

from __future__ import annotations

import logging
from typing import Annotated

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from arc.api.anime_schemas import AnimeSummary
from arc.api.anime_schemas import SearchPage as AnimeSearchPage
from arc.api.deps import CatalogDep, CurrentUser, SessionDep, SettingsDep
from arc.api.review_schemas import (
    CandidateOut,
    ConfirmRequest,
    ReviewItem,
    ReviewPage,
    ReviewSummary,
)
from arc.config import Settings
from arc.models import Anime, MediaFile, ReviewState
from arc.services.acquisition.reject import reject_download
from arc.services.catalog import (
    CATALOGUE_UNAVAILABLE,
    SourceUnavailable,
    list_status_for,
    upsert_summaries,
)
from arc.services.library.link import LinkError, link

log = logging.getLogger(__name__)

router = APIRouter(prefix="/api/review", tags=["review"])

ITEM_NOT_FOUND = "review item not found"
ALREADY_LINKED = "this file is already linked to an episode"
NOT_IGNORED = "only an ignored file can be reopened"

#: How many rows one page of the queue returns. The queue is meant to be
#: emptied, not paged through; a library that has hundreds pending has a
#: matcher problem, and the cap keeps a runaway one from being a 40 MB
#: response.
PAGE_LIMIT = 200

#: Same bounds as ``/api/anime/search``, so the review UI's search box and the
#: main one behave identically.
MIN_QUERY_LENGTH = 2
MAX_QUERY_LENGTH = 100


async def _pending_count(session: AsyncSession) -> int:
    total = await session.scalar(
        select(func.count())
        .select_from(MediaFile)
        .where(MediaFile.review_state == ReviewState.PENDING)
    )
    return int(total or 0)


async def _shows_for(session: AsyncSession, rows: list[MediaFile]) -> dict[int, Anime]:
    """Every anime named by every candidate of every row, in one query.

    Resolving them per candidate would be a round trip each, and a queue of
    fifty files with five candidates apiece is two hundred and fifty of them.
    """
    wanted: set[int] = set()
    for row in rows:
        wanted |= CandidateOut.anime_ids(row.match_candidates)
    if not wanted:
        return {}
    found = await session.scalars(select(Anime).where(Anime.id.in_(wanted)))
    return {anime.id: anime for anime in found.all()}


async def _item(session: AsyncSession, media_file: MediaFile, settings: Settings) -> ReviewItem:
    shows = await _shows_for(session, [media_file])
    return ReviewItem.build(media_file, data_dir=settings.data_dir, shows=shows)


async def _load(session: AsyncSession, media_file_id: int) -> MediaFile:
    media_file = await session.get(MediaFile, media_file_id)
    if media_file is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=ITEM_NOT_FOUND)
    return media_file


@router.get(
    "",
    response_model=ReviewPage,
    summary="The match-review queue (FR-L4, FR-L6)",
)
async def list_queue(
    user: CurrentUser,
    session: SessionDep,
    settings: SettingsDep,
    state: ReviewState = ReviewState.PENDING,
    limit: Annotated[int, Query(ge=1, le=PAGE_LIMIT)] = PAGE_LIMIT,
) -> ReviewPage:
    """Files in ``state``, newest first, plus the always-pending count."""
    rows = list(
        (
            await session.scalars(
                select(MediaFile)
                .where(MediaFile.review_state == state)
                .order_by(MediaFile.created_at.desc(), MediaFile.id.desc())
                .limit(limit)
            )
        ).all()
    )
    shows = await _shows_for(session, rows)
    return ReviewPage(
        items=[ReviewItem.build(row, data_dir=settings.data_dir, shows=shows) for row in rows],
        pending=await _pending_count(session),
    )


@router.get(
    "/summary",
    response_model=ReviewSummary,
    summary="How many files are waiting for review",
)
async def summary(user: CurrentUser, session: SessionDep) -> ReviewSummary:
    """One ``COUNT``. The client sidebar polls this, so it stays this small."""
    return ReviewSummary(pending=await _pending_count(session))


@router.get(
    "/{media_file_id}/search",
    response_model=AnimeSearchPage,
    summary="Search the catalogue for a title to confirm against (FR-L6)",
    responses={
        404: {"description": ITEM_NOT_FOUND},
        502: {"description": CATALOGUE_UNAVAILABLE},
    },
)
async def search_for_item(
    media_file_id: int,
    user: CurrentUser,
    session: SessionDep,
    catalog: CatalogDep,
    q: Annotated[str, Query(min_length=MIN_QUERY_LENGTH, max_length=MAX_QUERY_LENGTH)],
) -> AnimeSearchPage:
    """The same live search ``/api/anime/search`` does, hung off an item.

    Under the item's id rather than free-standing so the client does not have
    to hold two ideas at once, and so that a future audit line ("who searched
    what while resolving this file") has somewhere to go. The results are
    upserted on the way past exactly as they are in the main search, which is
    what gives each one the internal id ``confirm`` then takes.
    """
    await _load(session, media_file_id)
    try:
        found = await catalog.search(q)
    except SourceUnavailable as exc:
        log.warning("review search failed", extra={"query": q, "error": str(exc)})
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY, detail=CATALOGUE_UNAVAILABLE
        ) from exc

    rows = await upsert_summaries(session, found.results)
    await session.commit()
    statuses = await list_status_for(session, user_id=user.id, anime_ids=[row.id for row in rows])
    return AnimeSearchPage(
        results=[AnimeSummary.from_anime(row, statuses.get(row.id)) for row in rows],
        page=found.page,
        has_next=found.has_next,
    )


@router.post(
    "/{media_file_id}/confirm",
    response_model=ReviewItem,
    summary="Link a file to an episode by hand (FR-L6)",
    responses={404: {"description": ITEM_NOT_FOUND}, 409: {"description": ALREADY_LINKED}},
)
async def confirm(
    media_file_id: int,
    body: ConfirmRequest,
    user: CurrentUser,
    session: SessionDep,
    settings: SettingsDep,
) -> ReviewItem:
    """Set the show and the episode number, and move the episode to ``matched``.

    Only from a file that is not linked — see the module docstring for why
    that is the same rule as "only ``pending`` or ``ignored``".
    """
    media_file = await _load(session, media_file_id)
    if media_file.episode_id is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=ALREADY_LINKED)

    try:
        await link(
            session,
            media_file,
            anime_id=body.anime_id,
            episode_number=body.episode_number,
            review_state=ReviewState.CONFIRMED,
        )
    except LinkError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    # A person decided, so there is no longer a machine's confidence to report.
    # The candidates stay: they are the evidence that was on screen.
    media_file.match_confidence = None
    await session.commit()
    log.info(
        "review item confirmed",
        extra={
            "media_file_id": media_file.id,
            "user_id": user.id,
            "anime_id": body.anime_id,
            "episode_number": body.episode_number,
        },
    )
    return await _item(session, media_file, settings)


@router.post(
    "/{media_file_id}/ignore",
    response_model=ReviewItem,
    summary='Mark a file "not anime / ignore" (FR-L6)',
    responses={404: {"description": ITEM_NOT_FOUND}},
)
async def ignore(
    media_file_id: int,
    user: CurrentUser,
    session: SessionDep,
    settings: SettingsDep,
) -> ReviewItem:
    """Take the file out of the queue without linking it.

    The row stays — the file is still on disk, and re-creating it on the next
    scan only to ignore it again would be a queue that never empties.

    If the file is one **Arc downloaded**, ignoring it also says something
    about the *episode*: the release Arc picked was not it, and the episode has
    been sitting in ``matching`` waiting for this answer. It goes to
    ``unavailable``, which is both the sentence the show page shows and the
    state the daily retry looks for — see
    :mod:`arc.services.acquisition.reject`.
    """
    media_file = await _load(session, media_file_id)
    media_file.review_state = ReviewState.IGNORED
    episode = await reject_download(session, media_file, downloads_dir=settings.downloads_dir)
    await session.commit()
    log.info(
        "review item ignored",
        extra={
            "media_file_id": media_file.id,
            "user_id": user.id,
            "episode_unavailable": episode.id if episode is not None else None,
        },
    )
    return await _item(session, media_file, settings)


@router.post(
    "/{media_file_id}/reopen",
    response_model=ReviewItem,
    summary="Put an ignored file back in the queue (FR-L6)",
    responses={404: {"description": ITEM_NOT_FOUND}, 409: {"description": NOT_IGNORED}},
)
async def reopen(
    media_file_id: int,
    user: CurrentUser,
    session: SessionDep,
    settings: SettingsDep,
) -> ReviewItem:
    """Only from ``ignored``, and deliberately so.

    Reopening a *linked* file would have to decide what happens to the episode
    it is linked to — which may by then be transcoded, half-watched, and the
    reason somebody's MAL progress moved. That is an unlink operation with
    real consequences, and it belongs with the admin delete/re-fetch tools
    (FR-T4), not behind the same button as "I ignored this by mistake".
    """
    media_file = await _load(session, media_file_id)
    if media_file.review_state is not ReviewState.IGNORED:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=NOT_IGNORED)
    media_file.review_state = ReviewState.PENDING
    await session.commit()
    log.info("review item reopened", extra={"media_file_id": media_file.id, "user_id": user.id})
    return await _item(session, media_file, settings)


__all__ = [
    "ALREADY_LINKED",
    "ITEM_NOT_FOUND",
    "NOT_IGNORED",
    "PAGE_LIMIT",
    "router",
]
